### ISSUE TITLE
FontChoice cannot work on some characters.

### ISSUE THREAD

**Issue by Suyun514** [mintty/mintty#928]
OS: Windows 10 1903
MinTTY version: 3.0.6 (on Cygwin)

I set the font as `Cascadia Code` by the GUI and edited `$APPDATA/mintty/config` as
```plain
FontChoice=Han:2;
Font2=Microsoft Yahei
```

But some characters cannot displayed by `Microsoft Yahei`, such as `，`、`（`, `）`, `，` and `；` (maybe also `。`, it is displayed similarly by the two fonts).  They still displayed by the default Chinese font of the OS (`SimSun`).

![image](https://user-images.githubusercontent.com/52325399/66300430-dc44bf00-e927-11e9-9e7c-c5eab7fe9ea0.png)

**Comment by Suyun514** [comment]
The problem also cause while using `Consolas`, `Fira Code`, `Lucida Console`:

![image](https://user-images.githubusercontent.com/52325399/66303381-7bb88080-e92d-11e9-9e1f-bd31a0442f41.png)

![image](https://user-images.githubusercontent.com/52325399/66303415-8a9f3300-e92d-11e9-8b48-7a2bc3cfea99.png)

![image](https://user-images.githubusercontent.com/52325399/66303455-9db20300-e92d-11e9-9166-6e27a3d1915b.png)

**Comment by mintty** [comment]
The reason is that all characters mentioned above are not categorized as "Han" but as "Common" in the Unicode database: https://www.unicode.org/Public/UCD/latest/ucd/Scripts.txt
I acknowledge that this is not satisfying as Han punctuation is perceived as belonging to the Han domain with respect to font choice. However, the script classes used by mintty are technically derived from Unicode data. I'd welcome a proposal how to distinguish the affected characters in a convincing way.

**Comment by Suyun514** [comment]
Supporting font families instead of individual fonts may be a better way.

**Comment by mintty** [comment]
This is not the point of your problem.
The config maps [character ranges] -> [font (family?)].
The issue here is the coverage of some characters in the character range.

**Comment by Suyun514** [comment]
I am sorry.

Or offering a default `Script.txt` and allow users to edit it?

**Comment by BrianInglis** [comment]
Those characters appear to be in CJK/Halfwidth/Fullwidth ranges, so CJK if not Han, and common in the same sense as Han:
```
.  U+03000...U+0303F: CJK Symbols and Punctuation
　 U+03000 IDEOGRAPHIC SPACE
、 U+03001 IDEOGRAPHIC COMMA
。 U+03002 IDEOGRAPHIC FULL STOP
.  U+0FF00...U+0FFEF: Halfwidth and Fullwidth Forms
（ U+0FF08 FULLWIDTH LEFT PARENTHESIS
） U+0FF09 FULLWIDTH RIGHT PARENTHESIS
， U+0FF0C FULLWIDTH COMMA
； U+0FF1B FULLWIDTH SEMICOLON
```
as are characters in these ranges:
```
.  U+02E80...U+02EFF: CJK Radicals Supplement
.  U+02FF0...U+02FFF: Ideographic Description Characters
.  U+03000...U+0303F: CJK Symbols and Punctuation
.  U+031C0...U+031EF: CJK Strokes
.  U+03200...U+032FF: Enclosed CJK Letters and Months
.  U+03300...U+033FF: CJK Compatibility
.  U+03400...U+04DBF: CJK Unified Ideographs Extension A
.  U+04E00...U+09FFF: CJK Unified Ideographs
.  U+0F900...U+0FAFF: CJK Compatibility Ideographs
.  U+0FE30...U+0FE4F: CJK Compatibility Forms
.  U+0FF00...U+0FFEF: Halfwidth and Fullwidth Forms
.  U+1F200...U+1F2FF: Enclosed Ideographic Supplement
.  U+20000...U+2A6DF: CJK Unified Ideographs Extension B
.  U+2A700...U+2B73F: CJK Unified Ideographs Extension C
.  U+2B740...U+2B81F: CJK Unified Ideographs Extension D
.  U+2B820...U+2CEAF: CJK Unified Ideographs Extension E
.  U+2F800...U+2FA1F: CJK Compatibility Ideographs Supplement
```
Many characters in these ranges appear in the Unihan_*.txt database except the first two ranges for some reason. Similarly all CJK, East Asian. Ideograph, Unified and other files and properties really should be treated as part of this set. 
See [Unicode TR38/UAX38 PropertyExtension](https://www.unicode.org/reports/tr38/#PropertyExtension) and [Unicode TR44/UAX44 Unihan](https://www.unicode.org/reports/tr44/#Unihan) for how Unihan properties should be extended to non-Unihan characters. 
As commented, lack of comprehensive, reliable data sources, and/or project resources, with the large gamut of characters, means that properties are most complete for characters from Chinese sources, and may be incomplete for other characters.

**Comment by BrianInglis** [comment]
Darn, missed two more sets of Radicals: 
```
.  U+02F00...U+02FDF: Kangxi Radicals
.  U+0A490...U+0A4CF: Yi Radicals
```
See also [Unicode TR24/UAX24 Script Anomalies](http://www.unicode.org/reports/tr24/#Script_Anomalies) which explains why these characters have been left Common and suggests checking compatibility decompositions "NFKD normalized text, in which these characters decompose into sequences including the Han" and "association with East Asian character sets". 
They also suggest "Implementations that wish to have Script property values that are preserved over compatibility equivalence would tailor the Script property values for these characters." which implies supporting a Script.txt extension facility as suggested by the OP e.g. option `UnihanExtension=<unicode-ranges-list>`.

**Comment by mintty** [comment]
Thanks, Brian.
I can easily derive a "CJK" set with something like `egrep -e (CJK|Radical|Ideographic|width.*Forms) Blocks.txt`.
But this set overlaps with the scripts ranges as defined in `Scripts.txt` (see below, without "Common"), so how should they be related?
```
U+02E80...U+02EFF: CJK Radicals Supplement
	Han
U+02F00...U+02FDF: Kangxi Radicals
	Han
U+02FF0...U+02FFF: Ideographic Description Characters
	Han
U+03000...U+0303F: CJK Symbols and Punctuation
	Han, Hangul, Inherited
U+031C0...U+031EF: CJK Strokes
U+03200...U+032FF: Enclosed CJK Letters and Months
	Hangul, Katakana
U+03300...U+033FF: CJK Compatibility
	Katakana
U+03400...U+04DBF: CJK Unified Ideographs Extension A
	Han
U+04E00...U+09FFF: CJK Unified Ideographs
	Han
U+0A490...U+0A4CF: Yi Radicals
	Yi
U+0F900...U+0FAFF: CJK Compatibility Ideographs
	Han
U+0FE30...U+0FE4F: CJK Compatibility Forms
U+0FF00...U+0FFEF: Halfwidth and Fullwidth Forms
	Latin, Katakana, Hangul
U+1F200...U+1F2FF: Enclosed Ideographic Supplement
	Hiragana
U+20000...U+2A6DF: CJK Unified Ideographs Extension B
	Han
U+2A700...U+2B73F: CJK Unified Ideographs Extension C
	Han
U+2B740...U+2B81F: CJK Unified Ideographs Extension D
	Han
U+2B820...U+2CEAF: CJK Unified Ideographs Extension E
	Han
U+2F800...U+2FA1F: CJK Compatibility Ideographs Supplement
	Han
```

**Comment by BrianInglis** [comment]
As my previous post said, Unicode suggest treating them all as Unihan where that trumps compatibility, as in East Asian contexts. 
If characters in those blocks or scripts are used for purposes other than native communication, users are unlikely to have much expectation of rendering, as when I have used _SQUARE_ units of measure etc. characters (in the same block as _TELEGRAPH_ time digits+units) to squish headings and line up columns. 
I do think that glyphs should not be narrowed unless they are expected to overflow a character cell, as that can make them impossibly small to see clearly, even at normal font sizes (9+). 
In such cases, I have to launch the Windows/X GUI editor `gvim` instead of using the console editor `vim`. 
I use `grep|less|vim +/pattern` on my `unicode-symbols.txt` more than I would expect, as I keep it up to date with characters labelled in new Unicode releases.

**Comment by mintty** [comment]
That does not clearly respond to my concern about the intersection of ranges defined according to Blocks.txt vs. Scripts.txt. Would you suggest I should drop the "Han" and "Hangul" classes and cover all ranges above by "CJK"? What about Hiragana and Katakana then?

About narrowing, that's actually another issue, but your condition is exactly the background in mintty; only if a glyph is in fact detected to be wider than a character cell, auto-narrowing is applied.

**Comment by BrianInglis** [comment]
I take the Unicode suggestion to mean to treat the union of the classes as Unihan if in an East Asian locale; if not, either do that, or preserve compatibility if there is a reason to do so (I don't know what the issues are, perhaps important in a visual document preparation and rendering package?) 
I notice a large number of symbol glyphs in monospace fonts take more space than a character cell, but as they are followed by blanks, still appear aligned, although that could be an artifact of ClearType, Uniscribe, or other font rendering tweaks.

**Comment by Suyun514** [comment]
Supporting font families can **solve the problem**. And the point of my problem is also to **solve the problem** (instead of talking about character ranges, the better way is to give up using it). So why not use the solution?

**Comment by mintty** [comment]
What solution? If you claim a substantially different approach should be taken, then describe it.

**Comment by Suyun514** [comment]
Supporting font family instead of a fixed `script.txt`.
Users can set more than one fonts. If the current fonts can not displayed the character, the next one will be used.

**Comment by mintty** [comment]
So if the user's primary font contains a small subset of CJK, for example, it would be displayed using primary font, and other CJK characters would use different font. Resulting in inconsistent display style which I would not appreciate.

**Comment by Suyun514** [comment]
I have not seen any fonts like that, the fonts I seen have either English and CJK (such as `SimSun` and `Sarasa Mono`) or only English (`Consolas` and so soon). 
Also, some popular softwares like Visual Studio Code also use the solution.

**Comment by BrianInglis** [comment]
MS Windows _redefines_ font **family** to be one of `Decorative` (poster), `Modern` (monospace), `Roman` (proportional serif), `Script` (cursive handwriting), `Swiss` (proportional sans serif), `Dontcare` (generic). 
Of course, to be confusing and self-contradictory, MS also calls each of Arial Black, Arial Narrow, and Arial Rounded font **families**. 
MS Windows does not _support_ font **family** in the sense you probably mean. 
That is, if you mean **family**, as in the X Window system X11R3+ Logical Font Description specification in the next field after foundry: **family**. 
This is normally synonymous with the typographical term **typeface**, referring to the overall design characteristics and relative proportions of the glyphs, although some differentiate **typefaces** within families (like the varieties of Arial mentioned above; see below). 
Within which **family/typeface** fonts may vary across various weights, orientations/slants, widths, styles, spacings, character set registries and encodings, plus scripts, or blocks. 
See [survey about terms](http://www.thomasphinney.com/2009/04/font-terms-survey-results/).

**Comment by Suyun514** [comment]
Maybe my statement has some problems, "font family" is only stand for "multiple fonts" or "font fallback".

**Comment by mintty** [comment]
What you'd like is available on Windows level;
see e.g. https://github.com/mintty/mintty/issues/883#issuecomment-500468655 or https://github.com/mintty/mintty/issues/892#issuecomment-511666827.

Of course you could desire a similar configuration option on application level, with some problems, however:
* there is no design proposal (as I had asked to suggest)
* it may be complex to implement
* it may affect display performance

**Comment by mintty** [comment]
So the plan is to add a meta-script "CJK" which comprises some of the already defined ones, Han, Hangul, Katakana, Hiragana, Bopomofo, plus special ranges is suggested above (`egrep -e (CJK|Radical|Ideographic|width.*Forms) Blocks.txt`). Should the following ranges be included?
3190..319F; Kanbun
FF00..FFEF; Halfwidth and Fullwidth Forms (or only some of them?)
1D300..1D35F; Tai Xuan Jing Symbols
1D360..1D37F; Counting Rod Numerals
others?

**Comment by mintty** [comment]
I was a bit mislead by the initial report with respect to 、and 。which are actually already covered by the "Han" script configuration. Not sure why they don't appear differently; maybe simply the font fallback renders them from the same font (so using Yahei anyway, even in the unconfigured case).

Whatever, I have uploaded a patch that introduces a common meta-script "CJK" that should cover all of Han, Hangul, Katakana, Hiragana, Bopomofo, Kanbun, Halfwidth and Fullwidth unless already assigned to "Latin", and a number of CJK symbol ranges. Documentation pending. Feel free to test the repository version.

**Comment by mintty** [comment]
Released 3.1.0.
